### Homework 2

### Question 1: Comparing Two Population Means (Independent t-test)

### Hypotheses

- Null hypothesis ($H_0$): The mean increase in sleep is the same for both drugs, $\mu_1 - \mu_2 = 0$.
- Alternative hypothesis ($H_A$): The mean increase in sleep differs between the drugs, $\mu_1 - \mu_2 \ne 0$.

Because the question asks whether the means differ, this is a two-sided independent-samples t-test. Welch's version is used because it does not assume equal population variances.

The significance level is $\alpha = 0.05$.

In [ ]:
import numpy as np
from scipy import stats


sleep_drug_1 = np.array([0.7, -1.6, -0.2, -1.2, -0.1, 3.4, 3.7, 0.8, 0.0, 2.0])
sleep_drug_2 = np.array([1.9, 0.8, 1.1, 0.1, -0.1, 4.4, 5.5, 1.6, 4.6, 3.4])

alpha = 0.05
result = stats.ttest_ind(sleep_drug_1, sleep_drug_2, equal_var=False)
mean_difference = sleep_drug_1.mean() - sleep_drug_2.mean()

print(f"Drug 1 mean: {sleep_drug_1.mean():.2f} hours")
print(f"Drug 2 mean: {sleep_drug_2.mean():.2f} hours")
print(f"Mean difference (drug 1 - drug 2): {mean_difference:.2f} hours")
print(f"Welch t-statistic: {result.statistic:.3f}")
print(f"Two-sided p-value: {result.pvalue:.4f}")

if result.pvalue < alpha:
    print("Decision: Reject H0. The mean increases differ significantly.")
else:
    print("Decision: Fail to reject H0. There is not enough evidence of a significant difference.")

Drug 1 mean: 0.75 hours
Drug 2 mean: 2.33 hours
Mean difference (drug 1 - drug 2): -1.58 hours
Welch t-statistic: -1.861
Two-sided p-value: 0.0794
Decision: Fail to reject H0. There is not enough evidence of a significant difference.


### Results and conclusion

- The sample mean increase was 0.75 hours for drug 1 and 2.33 hours for drug 2. 
- Thus, drug 2 had a sample mean increase that was 1.58 hours larger than drug 1.

The Welch independent t-test produced $t = -1.861$ and a two-sided $p$-value of 0.0794. Since $p > 0.05$, we fail to reject $H_0$. There is not sufficient statistical evidence at the 5% significance level to conclude that the mean increase in sleep differs between the two drugs.

This conclusion does not prove that the drugs have identical effects; it means that this sample does not provide strong enough evidence of a difference at $\alpha = 0.05$.

### Question 2 ANOVA 
The iris dataset contains measurements for 50 flowers from each of three species of iris: 
setosa, versicolor, and virginica. 

### Context: 
- The iris dataset contains measurements for 50 flowers from each of three species of iris:setosa, versicolor, and virginica. 

### Question 2: 
- Is there a statistically significant difference in the mean sepal length (Sepal.Length) among the three different iris species (Species)? 
- If so, which groups differ?

In [6]:
import pandas as pd 
import numpy as np 
import seaborn as sns


iris_data = sns.load_dataset('iris')

In [7]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

alpha = 0.05

# One-way ANOVA: compare mean sepal length across the three species.
groups = [
    iris_data.loc[iris_data["species"] == species, "sepal_length"]
    for species in iris_data["species"].unique()
]
anova_result = stats.f_oneway(*groups)

print("Mean sepal length by species:")
print(iris_data.groupby("species")["sepal_length"].mean().round(3))
print(f"\nANOVA F-statistic: {anova_result.statistic:.3f}")
print(f"ANOVA p-value: {anova_result.pvalue:.6g}")

if anova_result.pvalue < alpha:
    print("\nDecision: Reject H0. At least one species mean differs.")

    tukey_result = pairwise_tukeyhsd(
        endog=iris_data["sepal_length"],
        groups=iris_data["species"],
        alpha=alpha,
    )
    print("\nTukey HSD post-hoc comparisons:")
    print(tukey_result)
else:
    print("\nDecision: Fail to reject H0. There is not enough evidence that the means differ.")

Mean sepal length by species:
species
setosa        5.006
versicolor    5.936
virginica     6.588
Name: sepal_length, dtype: float64

ANOVA F-statistic: 119.265
ANOVA p-value: 1.66967e-31

Decision: Reject H0. At least one species mean differs.

Tukey HSD post-hoc comparisons:
   Multiple Comparison of Means - Tukey HSD, FWER=0.05   
  group1     group2   meandiff p-adj lower  upper  reject
---------------------------------------------------------
    setosa versicolor     0.93   0.0 0.6862 1.1738   True
    setosa  virginica    1.582   0.0 1.3382 1.8258   True
versicolor  virginica    0.652   0.0 0.4082 0.8958   True
---------------------------------------------------------


### Results and conclusion

The mean sepal lengths are:

- **Setosa:** 5.006
- **Versicolor:** 5.936
- **Virginica:** 6.588

The one-way ANOVA produced F(2,147) = 119.265 with $p \approx 1.67 \times 10^{-31}$. Because $p < 0.05$, we reject $H_0$ and conclude that mean sepal length is not the same for all three iris species.

Tukey's HSD post-hoc test found that every pair differs significantly at the 5% level:

- Setosa has a significantly shorter mean sepal length than versicolor.
- Setosa has a significantly shorter mean sepal length than virginica.
- Versicolor has a significantly shorter mean sepal length than virginica.

Therefore, all three species differ significantly in mean sepal length, with virginica having the largest sample mean and setosa the smallest.

### Question 3 Chi-Square Test for Independence 


### Context: 
- The mtcars dataset gives performance data for 32 different cars. The variable cyl 
represents the number of cylinders (4, 6, or 8) and am represents the transmission type (0 = 
automatic, 1 = manual).  

### Question 3: 
- Is there a statistically significant association between the number of cylinders a car has 
and its transmission type? 

In [ ]:
import pandas as pd
from scipy import stats 

# Chi-Square Test for Independence between number of cylinders and transmission type in the mtcars dataset

cylinders = [6, 6, 4, 6, 8, 6, 8, 4, 4, 6, 6, 8, 8, 8, 8, 8,
            8, 4, 4, 4, 4, 8, 8, 8, 8, 4, 4, 4, 8, 8, 4]
transmission = [1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1]

mtcars = pd.DataFrame({"cyl": cylinders, "am": transmission})
contingency_table = pd.crosstab(mtcars["cyl"], mtcars["am"])
chi2_result = stats.chi2_contingency(contingency_table)

print("Contingency table (rows = cylinders, columns = transmission):")
print(contingency_table.rename(columns={0: "automatic", 1: "manual"}))
print("\nExpected counts:")
print(pd.DataFrame(
    chi2_result.expected_freq,
    index=contingency_table.index,
    columns=["automatic", "manual"],
).round(2))
print(f"\nChi-square statistic: {chi2_result.statistic:.3f}")
print(f"Degrees of freedom: {chi2_result.dof}")
print(f"p-value: {chi2_result.pvalue:.6f}")

if chi2_result.pvalue < 0.05:
    print("\nDecision: Reject H0. Cylinder count and transmission type are associated.")
else:
    print("\nDecision: Fail to reject H0. There is not enough evidence of an association.")

Contingency table (rows = cylinders, columns = transmission):
am   automatic  manual
cyl                   
4            3       8
6            4       2
8           12       2

Expected counts:
     automatic  manual
cyl                   
4         6.74    4.26
6         3.68    2.32
8         8.58    5.42

Chi-square statistic: 8.958
Degrees of freedom: 2
p-value: 0.011343

Decision: Reject H0. Cylinder count and transmission type are associated.


### Results and conclusion

- Null hypothesis ($H_0$): The number of cylinders and transmission type are independent.
- Alternative hypothesis ($H_A$): The number of cylinders and transmission type are associated.

The contingency table is:

| Cylinders | Automatic | Manual |
|---:|---:|---:|
| 4 | 3 | 8 |
| 6 | 4 | 2 |
| 8 | 12 | 2 |

The chi-square test produced $\chi^2(2) = 8.958$ with a $p$-value of 0.011343. Since $p < 0.05$, we reject $H_0$ and conclude that the number of cylinders and transmission type are statistically significantly associated.

In this sample, manual transmissions are more common among 4-cylinder cars, while automatic transmissions are especially common among 8-cylinder cars.

### Question 4 Chi-Square Goodness-of-Fit Test 


### Context: 
- An economist claims that among commuters in a large city, the preferred mode of 
transport is distributed as follows: 50% use public transit, 30% drive their own car, and 20% use 
other methods (cycling, walking, etc.). A random sample of 300 commuters is taken, and their 
choices are recorded: 135 use public transit, 95 drive their own car, and 70 use other methods.   

### Question 4: 
-  At the alpha=0.01 significance level, does the sample data provide evidence to reject the 
economist's claim about the distribution of commuter transport preferences? 

In [12]:
import numpy as np
from scipy import stats

observed_counts = np.array([135, 95, 70])
claimed_proportions = np.array([0.50, 0.30, 0.20])
sample_size = observed_counts.sum()
expected_counts = sample_size * claimed_proportions
alpha = 0.01

gof_result = stats.chisquare(
    f_obs=observed_counts,
    f_exp=expected_counts,
)

categories = ["Public transit", "Own car", "Other methods"]
print("Observed counts:", dict(zip(categories, observed_counts)))
print("Expected counts:", dict(zip(categories, expected_counts)))
print(f"\nChi-square statistic: {gof_result.statistic:.3f}")
print(f"Degrees of freedom: {len(observed_counts) - 1}")
print(f"p-value: {gof_result.pvalue:.6f}")

if gof_result.pvalue < alpha:
    print("\nDecision: Reject H0. The sample distribution differs from the claimed distribution.")
else:
    print("\nDecision: Fail to reject H0. There is not enough evidence to reject the claimed distribution.")

Observed counts: {'Public transit': 135, 'Own car': 95, 'Other methods': 70}
Expected counts: {'Public transit': 150.0, 'Own car': 90.0, 'Other methods': 60.0}

Chi-square statistic: 3.444
Degrees of freedom: 2
p-value: 0.178669

Decision: Fail to reject H0. There is not enough evidence to reject the claimed distribution.


### Results and conclusion

- Null hypothesis ($H_0$): The commuter preferences follow the claimed distribution $(0.50, 0.30, 0.20)$.
- Alternative hypothesis ($H_A$): The commuter preferences do not follow the claimed distribution.

For a sample of 300 commuters, the expected counts are 150 public transit users, 90 own-car drivers, and 60 users of other methods. The observed counts are 135, 95, and 70, respectively.

The chi-square goodness-of-fit test produced $\chi^2(2) = 3.444$ with a $p$-value of **0.178669**. Because $p > \alpha=0.01$, we **fail to reject $H_0$**.

Therefore, at the 1% significance level, the sample does not provide sufficient evidence to reject the economist's claim about the distribution of commuter transport preferences.

### Question 5 Nonparametric Testing 


### Context: 
- Revisit the sleep dataset from Question 1. One key assumption of the t-test is that the data 
for each group is approximately normally distributed, especially with small sample sizes (n=10 for 
each group).   

### Question 5: 
-  If we are unwilling to assume normality for the extra sleep data, what is an appropriate 
nonparametric alternative to the t-test? Perform this test and compare your conclusion to the one 
from Question 1. 

In [13]:
mann_whitney_result = stats.mannwhitneyu(
    sleep_drug_1,
    sleep_drug_2,
    alternative="two-sided",
)

print(f"Drug 1 median: {np.median(sleep_drug_1):.2f} hours")
print(f"Drug 2 median: {np.median(sleep_drug_2):.2f} hours")
print(f"Mann-Whitney U statistic: {mann_whitney_result.statistic:.1f}")
print(f"Two-sided p-value: {mann_whitney_result.pvalue:.4f}")

if mann_whitney_result.pvalue < 0.05:
    print("Decision: Reject H0. The two groups differ significantly.")
else:
    print("Decision: Fail to reject H0. There is not enough evidence of a significant difference.")

Drug 1 median: 0.35 hours
Drug 2 median: 1.75 hours
Mann-Whitney U statistic: 25.5
Two-sided p-value: 0.0693
Decision: Fail to reject H0. There is not enough evidence of a significant difference.


### Results and comparison with Question 1

The appropriate nonparametric alternative to the independent-samples t-test is the **two-sided Mann-Whitney U test**. It does not require normally distributed data and compares the rank distributions of the two independent groups.

- Null hypothesis ($H_0$): The two groups come from the same distribution.
- Alternative hypothesis ($H_A$): The two groups come from different distributions.

The median increase was 0.35 hours for drug 1 and 1.75 hours for drug 2. The Mann-Whitney test produced $U=25.5$ with a two-sided $p$-value of **0.0693**. Since $p>0.05$, we fail to reject $H_0$.

This agrees with Question 1: the Welch t-test also failed to find a statistically significant difference ($p=0.0794$). Therefore, whether we compare means with the t-test or rank distributions with the Mann-Whitney test, this sample does not provide sufficient evidence of a difference between the drugs at the 5% significance level.